src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -112,8 +127,55 @@ private function completeDefinition($id, Definition $definition)
 
             if ($argumentExists) {
                 $definition->replaceArgument($index, $value);
-            } else {
-                $definition->addArgument($value);
+
+                continue;
+            }
+
+            $definition->addArgument($value);
+        }
+    }
+
+    /**
+     * Autowires setters.
+     *
+     * @param string           $id
+     * @param Definition       $definition
+     * @param \ReflectionClass $reflectionClass
+     */
+    private function autowireSetters($id, Definition $definition, \ReflectionClass $reflectionClass)
+    {
+        $methodsCalled = array();
+        foreach ($definition->getMethodCalls() as $methodCall) {
+            $methodsCalled[$methodCall[0]] = true;
+        }
+
+        foreach ($reflectionClass->getMethods(\ReflectionMethod::IS_PUBLIC) as $reflectionMethod) {
+            $name = $reflectionMethod->getName();
+            if (isset($methodsCalled[$name]) || $reflectionMethod->isStatic() || 0 !== strpos($name, 'set') || 1 !== $reflectionMethod->getNumberOfParameters()) {
+                continue;
+            }
+
+            $parameter = $reflectionMethod->getParameters()[0];
+            try {
+                if (!$typeHint = $parameter->getClass()) {
+                    continue;
+                }
+
+                if (null === $this->types) {
+                    $this->populateAvailableTypes();
+                }
+
+                if (isset($this->types[$typeHint->name])) {
+                    $value = new Reference($this->types[$typeHint->name]);
```

Comment:
It's ignored because the `try`/`catch` block. I don't take the default value because in such case  (`setFoo(Foo $foo = null)`), calling the setter is useless.